packages/calypso-e2e/src/lib/blocks/block-flows/ai-assistant.ts

```diff
@@ -0,0 +1,158 @@
+		await block.waitFor( { state: 'detached' } );
+		this.generatedText = await editorCanvas.locator( '.is-selected' ).innerText();
+	}
+
+	/**
+	 * Clicks on the toolbar button, then in the dropdown that appears, clicks an option.
+	 *
+	 * @param {string} buttonName Name of the button to click.
+	 * @param {string} optionName Name of the option in the dropdown to click.
+	 * @param {EditorContext} context The current context for the editor at the point of
+	 * test execution.
+	 */
+	private async clickToolbarButton(
+		buttonName: string,
+		optionName: string,
+		context: EditorContext
+	) {
+		// Block-based and non-block-based themes strike again.
+		// For block-based themes (eg. Twenty Twenty-Three) a given block's toolbar
+		// lives "outside" of the DOM tree returned by `getEditorCanvas`.
+		// For non-block-based themes (eg. Twenty Twenty-One) a block's toolbar
+		// lives in the same DOM tree as the locator returned by `getEditorCanvas`.
+		// However, the locator returned by `getEditorParent` has the toolbar inside
+		// the DOM tree in both cases.
+		const editorParent = await context.editorPage.getEditorParent();
+
+		const toolbar = editorParent.getByRole( 'toolbar', { name: 'Block tools' } );
+		await toolbar.waitFor();
+		// Yes, yuck, I know. But Playwright often tries to act on the buttons sooner
+		// than it should, so we need to wait for the bounding box of the element here
+		// to be stable before attempting to click on the toolbar.
+		const elementHandle = await toolbar.elementHandle();
+		await elementHandle?.waitForElementState( 'stable' );
+
+		const button = toolbar.getByRole( 'button', { name: buttonName } );
+		await button.waitFor();
+		await button.click();
+
+		const option = editorParent.getByRole( 'menuitem', { name: optionName } );
+		await option.waitFor();
+		await option.click();
+	}
+
+	/**
+	 * Clicks on a button on the block.
+	 *
+	 * @param {Locator} block Locator to the block.
+	 * @param {string}name Name of the button to click.
+	 */
+	private async clickButtonOnBlock( block: Locator, name: string ) {
+		await block.getByRole( 'button', { name: name } ).click();
+	}
+
+	/**
+	 * Enters the query into the input on the block.
+	 *
+	 * @param {Locator} block Locator to the block.
+	 */
+	private async enterInput( block: Locator ) {
+		await block
+			.getByRole( 'textbox', { name: 'Ask Jetpack AI' } )
+			.fill( this.configurationData.query );
+	}
+
+	/**
+	 * Helper method to wait for the query to complete.
+	 *
+	 * @param {Locator} block Locator to the block.
+	 */
+	private async waitForQuery( block: Locator ) {
+		await Promise.all( [
+			block
+				.getByRole( 'button', { name: 'Stop request' } )
+				.waitFor( { state: 'detached', timeout: 15 * 1000 } ),
+			block
+				.getByRole( 'textbox', { name: 'Ask Jetpack AI', disabled: true } )
+				.waitFor( { state: 'detached', timeout: 15 * 1000 } ),
+			block.locator( '.components-spinner' ).waitFor( { state: 'detached' } ),
+		] );
```

Comment:
**Question (non-blocking):** Out of curiosity, why all three? These all seem to happen at once from everything I can tell, so I think we should be able to just key off of one, right? I think that would decrease the surface area of potential breaking DOM changes/flakiness, etc. WDYT?

Reply:
Good and valid question.

During development I consistently ran into issues with the test running under Jetpack AT environment racing past the waits. 

The first wait `Stop request` can probably be removed and instead we can rely on either the second or third waits.

Reply:
Ah sure, it sounds like maybe they don't always reliably end at the same time! Thanks for explaining! 😄